go/tasks/pluginmachinery/flytek8s/pod_helper.go

```diff
@@ -150,8 +155,22 @@ func BuildPodWithSpec(podTemplate *v1.PodTemplate, podSpec *v1.PodSpec) (*v1.Pod
 			return nil, err
 		}
 
-		basePodSpec.Containers = podSpec.Containers
-
+		var mergedContainers []v1.Container
+		for _, container := range podSpec.Containers {
+			for _, templateContainer := range basePodSpec.Containers {
+				if PrimaryContainerNameMatches(container.Name, templateContainer.Name, primaryContainerName) {
+					newContainer := templateContainer.DeepCopy()
+					err := mergo.Merge(newContainer, container)
+					if err != nil {
+						return nil, err
+					}
+					mergedContainers = append(mergedContainers, *newContainer)
+				} else {
+					mergedContainers = append(mergedContainers, container)
+				}
+			}
+		}
```

Comment:
May need to double-check the loop and container logic here. I think in this scenario the case where the template containers:
```
["foo", "bar", "default"]
```
and the podSpec from flyte contains containers:
```
["baz", "default"]
```
the resulting container set will be
```
["foo", "bar", "default"]
```
when it should be:
```
["baz", "default"]
```
with the configuration of default using the podTemplate configuration. Am I thinking about this correctly? Or should the result be:
```
["foo", "bar", "baz", "default"]
```
where any container in the template is either the base configuration for a pod container or added?

Reply:
In the behavior that you were expecting, do `foo` and `bar` from the template also get merged into the primary container `baz`? Or the default? (I think I don't have a good example to think through for why we'd want this behavior, since the use case I'm concerned with is more limited)

Reply:
Yeah, so I don't think we should ever merge anything other than the "default" or "primary-container" in the template with the default podSepc container.

In this example I think the "foo" and "bar" containers from the template will be added to the podSpec, without merging with anything. The only use-case I can think of for this functionality is the ability to add a container to every pod that is started by Flyte. For example if we had the PodTemplate:
```
apiVersion: v1
kind: PodTemplate
metadata:
  name: flyte-default-template
  namespace: flyte
template:
  spec:
    containers:
      - name: primary-container-name
        volumeMounts:
          - name: foo
            mountPath: /foo
      - name: noop
        image: docker.io/rwgrim/docker-noop
```
We would obviously use the `primary-container-name` container as the base configuration for all of the primary containers. The `noop` container, while admittedly non-useful in this example, could be something more interesting. The question is, should we either:
(1) never use this container
(2) add the container to the Pod

I think for simplicity option 1 is better, I think we would have to change logic of how the container plugin executes with option 2. Also, this is something we can revisit in the future if we want.

Reply:
Got it thank you, that makes a ton of sense! I was definitely thinking of option 2 because I wasn't sure if we would be okay with not running it, but as I said I can't think of a clear use case for adding multiple to the pod like this, so I'm good with option 1 as it also satisfies what I'm looking for with this change!